### PR TITLE
[Core] Embedded skin utility enhancements

### PR DESCRIPTION
**Pull request by rubenzorrilla**
Before, we used to do the interpolation of an embedded variable in the intersections skin by searching each node one by one and interpolating from the parent element nodal values. Even though this is fine for continuous level-set approaches, it is not for discontinuous case.

This PR, modifies the interpolation to use the modified shape function values calculated in the intersected edges. This must not affect the results of the continuous approach but allows to consider the discontinuous with a unique implementation.

### CODE REVIEW

File: kratos/utilities/embedded_skin_utility.cpp [hunk 1]
```diff
@@ -277,10 +335,87 @@ namespace Kratos
             case GeometryData::KratosGeometryType::Kratos_Tetrahedra3D4:
                 return Kratos::make_shared<DivideTetrahedra3D4>(rGeometry, rNodalDistances);
             default:
-                KRATOS_ERROR << "Asking for a non-implemented modified shape functions geometry.";
+                KRATOS_ERROR << "Asking for a non-implemented divide geometry utility.";
         }
     }
 
+    template<std::size_t TDim>
+    ModifiedShapeFunctions::UniquePointer EmbeddedSkinUtility<TDim>::SetModifiedShapeFunctionsPointer(
```
Comment (RiccardoRossi) [inline kratos/utilities/embedded_skin_utility.cpp]: create?

File: kratos/utilities/embedded_skin_utility.h [hunk 2]
```diff
@@ -277,13 +350,74 @@ class KRATOS_API(KRATOS_CORE) EmbeddedSkinUtility
...
+     * From the split pattern of an intersected element, this method
+     * creates and returns a pointer to a new skin condition.
+     * @param rOriginGeometryType GeometryType of the condition to be created
+     * @param rNewNodesArray array containing pointers to the nodes that will
+     * conform the condition
+     * @param rConditionId new condition identifier
+     * @param pConditionProperties pointer to the new condition properties
+     * @return Condition::Pointer pointer to a new skin condition
+     */
+    Condition::Pointer SetNewConditionPointer(
```
Comment (RiccardoRossi) [inline kratos/utilities/embedded_skin_utility.h]: create?

File: kratos/utilities/embedded_skin_utility.cpp [hunk 3]
```diff
@@ -277,10 +335,87 @@ namespace Kratos
...
+        const Geometry<Node<3>>::Pointer pGeometry,
+        const Vector& rNodalDistances)
+    {
+        // Get the geometry type
+        const GeometryData::KratosGeometryType geometry_type = pGeometry->GetGeometryType();
+
+        // Return the modified shape functions utility
+        if (mLevelSetType == "continuous"){
+            switch (geometry_type){
```
Comment (RiccardoRossi) [inline kratos/utilities/embedded_skin_utility.cpp]: very slow

File: kratos/utilities/embedded_skin_utility.h [hunk 4]
```diff
@@ -77,10 +85,12 @@ class KRATOS_API(KRATOS_CORE) EmbeddedSkinUtility
     EmbeddedSkinUtility(
         ModelPart &rModelPart,
         ModelPart &rSkinModelPart,
-        const std::string LevelSetType = "continuous") :
+        const std::string LevelSetType = "continuous",
```
Comment (RiccardoRossi) [inline kratos/utilities/embedded_skin_utility.h:98]: do not use strings

File: kratos/utilities/embedded_skin_utility.cpp [hunk 5]
```diff
@@ -205,6 +241,28 @@ namespace Kratos
...
+        const unsigned int &rConditionId,
+        const Properties::Pointer pConditionProperties)
+    {
+        const auto &r_condition = KratosComponents<Condition>::Get(this->GetConditionType());
+        auto p_new_geom = this->SetNewConditionGeometry(rOriginGeometryType, rNewNodesArray);
+        return r_condition.Create(rConditionId, p_new_geom, pConditionProperties);
+    }
+
+    template<>
+    const std::string EmbeddedSkinUtility<2>::GetConditionType() const {
```
Comment (RiccardoRossi) [inline kratos/utilities/embedded_skin_utility.cpp]: use kratos_components only once

File: kratos/utilities/embedded_skin_utility.cpp (hunk 1 above)
Comment (rubenzorrilla) [reply kratos/utilities/embedded_skin_utility.cpp]: Renamed to `pCreateModifiedShapeFunctions`.

File: kratos/utilities/embedded_skin_utility.h (hunk 2 above)
Comment (rubenzorrilla) [reply kratos/utilities/embedded_skin_utility.h]: Renamed to `pCreateNewCondition`. Same has been done with the `pCreateNewConditionGeometry`.

File: kratos/utilities/embedded_skin_utility.cpp (hunk 3 above)
Comment (rubenzorrilla) [reply kratos/utilities/embedded_skin_utility.cpp]: Changed to `enum`

File: kratos/utilities/embedded_skin_utility.cpp (hunk 5 above)
Comment (rubenzorrilla) [reply kratos/utilities/embedded_skin_utility.cpp]: Done. Now a member reference to the prototype condition is used.

File: kratos/utilities/embedded_skin_utility.h (hunk 4 above)
Comment (rubenzorrilla) [reply kratos/utilities/embedded_skin_utility.h:98]: An `enum` type is used.

**Review by RiccardoRossi — APPROVED** [review]
+1

**Comment by loumalouomega** [comment]
My compilation is broken:

~~~sh
/usr/bin/ld: CMakeFiles/KratosCore.dir/tests/cpp_tests/utilities/test_embedded_skin_utility.cpp.o: en la función `Kratos::Testing::TestEmbeddedSkinUtility2D::TestFunction()':
test_embedded_skin_utility.cpp:(.text+0x2f6f): referencia a `Kratos::EmbeddedSkinUtility<2ul>::GetConditionType[abi:cxx11]()' sin definir
collect2: error: ld returned 1 exit status
make[2]: *** [kratos/CMakeFiles/KratosCore.dir/build.make:3148: kratos/libKratosCore.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:151: kratos/CMakeFiles/KratosCore.dir/all] Error 2
~~~

**Pull request merged**